common/src/main/java/io/pravega/common/util/CompositeByteArraySegment.java

```diff
@@ -0,0 +1,273 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.common.util;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Iterators;
+import io.pravega.common.Exceptions;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.SequenceInputStream;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import lombok.Getter;
+import lombok.NonNull;
+
+/**
+ * A composite, index-based array-like structure that is made up of one or more individual arrays of equal size. Each
+ * component array maps to a contiguous offset range and is only allocated when the first index within its range needs
+ * to be set (if unallocated, any index within its range will have a value of 0).
+ */
+public class CompositeByteArraySegment implements CompositeArrayView {
```

Comment:
Why aren't we using Netty ByteBufs ? There is already lot of support for composition and they support direct memory?